core/src/main/scala/kafka/server/ClientQuotaManager.scala

```diff
@@ -139,11 +140,21 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
    * we need to add a delay of X to W such that O * W / (W + X) = T.
    * Solving for X, we get X = (O - T)/T * W.
    */
-  private def throttleTime(metricValue: Double, config: MetricConfig): Int = {
+  private def throttleTime(clientMetric: KafkaMetric, config: MetricConfig): Int = {
+    // Casting to Rate because we only use Rate in Quota computation
+    val rateMetric: Rate = measurableAsRate(clientMetric.metricName(), clientMetric.measurable())
     val quota = config.quota()
-    val difference = metricValue - quota.bound
-    val time = difference / quota.bound * config.timeWindowMs() * config.samples()
-    time.round.toInt
+    val difference = clientMetric.value() - quota.bound
+    // Use the precise window used by the rate calculation
+    val throttleTimeMs = difference / quota.bound * rateMetric.windowSize(config, time.milliseconds())
+    throttleTimeMs.round.toInt
+  }
+
+  private def measurableAsRate(name: MetricName, measurable: Measurable): Rate = {
+    if (! measurable.isInstanceOf[Rate])
+      throw new IllegalArgumentException("Metric " + name + " is not a Rate metric")
+
+    measurable.asInstanceOf[Rate]
```

Comment:
Thanks for making this change. I'd suggest the more idiomatic with a few tweaks:

```
measurable match {
  case r: Rate => r
  case _ => throw new IllegalArgumentException(s"Metric $name is not a Rate metric, value $measurable")
```